Quorum/ipfs_validator.py

```diff
@@ -0,0 +1,69 @@
+from Quorum.utils.chain_enum import Chain
+import Quorum.config as config
+from Quorum.apis.block_explorers.chains_api import ChainAPI
+from Quorum.apis.llms.claude_conversation import ClaudeConversation
+
+import requests
+import argparse
+import re
+from pathlib import Path
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description='Compare ipfs with actual payload.')
+    parser.add_argument('--proposal_id', type=int, help='The id of the proposal.')
+    parser.add_argument('--chain', type=str, choices=[chain.value for chain in Chain], help='Blockchain chain.')
+    parser.add_argument('--proposal_address', type=str, help='Ethereum proposal address.')
+    parser.add_argument('--prompt_template', type=Path, default=Path(__file__).parent / 'ipfs_validation_prompt.txt',
+                        help='Optional other template for prompting the LLM.')
+    
+    args = parser.parse_args()
+    if not args.prompt_template.exists():
+        raise argparse.ArgumentError(f'Could not find prompts at {args.prompt_template}.')
+
+    return args
+
+
+def get_raw_ipfs(proposal_id: int) -> str:
+    resp = requests.get(f'https://vote.onaave.com/proposal/?proposalId={proposal_id}')
+    resp.raise_for_status()
+    voting_ui_html = resp.content.decode()
+
+    # Somewhere in the HTML there's a js script that contains a json with the following field:
+    # "ipfsHash": "<ipfs_hash>". We want to get the ipfs_hash to construct the link to the raw ipfs.
+    # The json also has \\" to mark these are special characters.
+    # ipfs_hash = voting_ui_html.replace('\\', '').split('"ipfsHash":')[1].split(',')[0].replace('"', '')
+    match = re.search(r'\\"ipfsHash\\":\s*\\"(.*?)\\"', voting_ui_html)
+    if not match:
+        raise ValueError('ipfsHash not found in the HTML.')
+    
+    ipfs_hash = match.group(1)
+    
+    resp = requests.get(f'https://ipfs.io/ipfs/{ipfs_hash}')
+    resp.raise_for_status()
+    return resp.content.decode()
+
+
+def main():
+    args = parse_args()
+
+    block_explorer = ChainAPI(args.chain)
+    source_codes = block_explorer.get_source_code(args.proposal_address)
+    payload = '\n'.join(source_codes[0].file_content)
+
+    ipfs = get_raw_ipfs(160)
+
+    with open(args.prompt_template) as f:
+        all_prompt_templates = f.read()
+    
+    prompts = [p.strip().replace('<ipfs>', ipfs).replace('<payload>', payload) for p in all_prompt_templates.split('<---next_prompt--->')]
```

Comment:
Just save the prompts under separate files with appropriate names, allowing you to load each prompt whenever needed.

Reply:
I'm actually not sure how Itay would prefer it and since we have multiple prompts just for one proposal type I didn't want to start messing with files. I'll ask Itay later what would be his best user experience since he's going to have tinker with this tool a bit more I assume.